server/src/test/java/org/elasticsearch/snapshots/mockstore/MockEventuallyConsistentRepository.java

```diff
@@ -0,0 +1,288 @@
+    private static List<BlobStoreAction> consistentView(List<BlobStoreAction> actions) {
+        final Map<String, BlobStoreAction> lastActions = new HashMap<>();
+        for (BlobStoreAction action : actions) {
+            if (action.operation == Operation.PUT) {
+                lastActions.put(action.path, action);
+            } else if (action.operation == Operation.DELETE) {
+                lastActions.remove(action.path);
+            }
+        }
+        return List.copyOf(lastActions.values());
+    }
+
+    @Override
+    protected void assertSnapshotOrGenericThread() {
+        // eliminate thread name check as we create repo in the test thread
+    }
+
+    @Override
+    protected BlobStore createBlobStore() {
+        return new MockBlobStore();
+    }
+
+    /**
+     * Context that must be shared between all instances of {@link MockEventuallyConsistentRepository} in a test run.
+     */
+    public static final class Context {
+
+        private final List<BlobStoreAction> actions = new ArrayList<>();
+
+        /**
+         * Force the repository into a consistent end state so that its eventual state can be examined.
+         */
+        public void forceConsistent() {
+            synchronized (actions) {
+                final List<BlobStoreAction> consistentActions = consistentView(actions);
+                actions.clear();
+                actions.addAll(consistentActions);
+            }
+        }
+    }
+
+    private enum Operation {
+        PUT, GET, DELETE
+    }
+
+    private static final class BlobStoreAction {
+
+        private final Operation operation;
+
+        @Nullable
+        private final byte[] data;
+
+        private final String path;
+
+        private BlobStoreAction(Operation operation, String path, byte[] data) {
+            this.operation = operation;
+            this.path = path;
+            this.data = data;
+        }
+
+        private BlobStoreAction(Operation operation, String path) {
+            this(operation, path, null);
+        }
+    }
+
+    private class MockBlobStore implements BlobStore {
+
+        private AtomicBoolean closed = new AtomicBoolean(false);
+
+        @Override
+        public BlobContainer blobContainer(BlobPath path) {
+            return new MockBlobContainer(path);
+        }
+
+        @Override
+        public void close() {
+            closed.set(true);
+        }
+
+        private void ensureNotClosed() {
+            if (closed.get()) {
+                throw new AssertionError("Blobstore is closed already");
+            }
+        }
+
+        private class MockBlobContainer implements BlobContainer {
+
+            private final BlobPath path;
+
+            MockBlobContainer(BlobPath path) {
+                this.path = path;
+            }
+
+            @Override
+            public BlobPath path() {
+                return path;
+            }
+
+            @Override
+            public boolean blobExists(String blobName) {
```

Comment:
Not for this PR, but AFAICS we can get rid of this method on BlobContainer. We only use it in BlobStoreRepository to check existence of `master.dat` but we could as well just read that file, it's super small

Reply:
Yea you're right. The main reason I kept it was actually that Azure used it as well for some magic with the secondary dns failover but obviously that can become an implementation detail there :) I put it on my list.

Reply:
https://github.com/elastic/elasticsearch/pull/44472